src/odemis/gui/cont/stream_bar.py

```diff
@@ -2185,3 +2186,97 @@ def clear(self, clear_model=True):
+    The only role of the controller is to display the streams in the
+    acquired view when the feature is selected.
+    """
+
+    def __init__(self, tab_data, feature_view, *args, **kwargs):
+        """
+        feature_view (StreamView): the view to show the feature streams
+        """
+        super().__init__(tab_data, *args, **kwargs)
+        self._feature_view = feature_view
+
+        tab_data.main.currentFeature.subscribe(self._on_current_feature_changes)
+
+    def showFeatureStream(self, stream) -> StreamController:
+        """
+        Shows an Feature stream (in the Acquired view)
+        Must be run in the main GUI thread.
+        """
+        self._feature_view.addStream(stream)
+        sc = self._add_stream_cont(stream, show_panel=True, static=self.static_mode,
+                                   view=self._feature_view)
+        return sc
+
+    @call_in_wx_main
+    def _on_current_feature_changes(self, feature):
+        """
+        Handle switching the acquired streams appropriate to the current feature
+        :param feature: (CryoFeature or None) the newly selected current feature
+        """
+        self.clear_feature_streams()
+        # show the feature streams on the acquired view
+
+        acquired_streams = []
+        if feature:
+            if feature.reference_image is not None:
+                acquired_streams = data_to_static_streams([feature.reference_image])
+            for stream in acquired_streams:
+                self.showFeatureStream(stream)
+                self.stream = stream # should only ever be 1 stream
+        # refit the selected feature in the acquired view
+        self._view_controller.viewports[3].canvas.fit_view_to_content()
+
+    def clear_feature_streams(self):
+        """
+        Remove from display all feature streams (but leave the overview and live streams)
+        But DO NOT REMOVE the streams from the model
+        """
+        # Remove the panels, and indirectly it will clear the view
+        v = self._feature_view
+        for sc in self.stream_controllers.copy():
+            if not isinstance(sc.stream, StaticSEMStream):
+                logging.warning("Unexpected non static stream: %s", sc.stream)
+                continue
+
+            self._stream_bar.remove_stream_panel(sc.stream_panel)
+            if hasattr(v, "removeStream"):
+                v.removeStream(sc.stream)
+            if sc in self.stream_controllers:
+                self.stream_controllers.remove(sc)
+
+        self.stream = None
+        self._stream_bar.fit_streams()
+
+        # Force a check of what can be garbage collected, as some of the streams
+        # could be quite big, that will help to reduce memory pressure.
+        gc.collect()
+
+    def clear(self, clear_model=True):
+        """
+        Remove all the streams, from the GUI (view, stream panels)
+        Must be called in the main GUI thread.
```

Comment:
I thought if something should be called in the main GUI thread, you should decorate it with `@call_in_wx_main`. But I see this comment comes back in some other places in Odemis and the decorator is not used in any of those places. Do you know why that's the case?

Reply:
I'm not sure either, but this it's the same in the above CryoAcquiredStreamController https://github.com/patrickcleeve2/odemis/blob/fbbd837d1c44c45c45ae077ac1b280320349668e/src/odemis/gui/cont/stream_bar.py#L2161 and in base StreamController: https://github.com/patrickcleeve2/odemis/blob/c1bb04bf7d87ab1e7f28d06a4a9a0fb12d207918/src/odemis/gui/cont/stream_bar.py#L900
My assumption is that the comment isn't actually true, or whatever calls it has the decorator.

Reply:
If it says "Must be called in the main GUI thread" that means the function is *not* protected (by the decorator). So the *caller* has to either be certain to be running in the main thread (ie, it's a GUI callback, or it's decorated), or it should call this function with `wx.CallLater()`.

It's useful to do it this way because the `@call_in_wx_main` decorator causes the call to be asynchronous, which means, it's slow, the control flow is hard to grasp, and especially it's not possible to get the return value.